### ISSUE TITLE
Is it possible to bypass articles that are not found?

### ISSUE THREAD

**Issue by DataAddict42** [openzim/mwoffliner#1976]
Is there any option to ignore or bypass articles that result in a 404? For example I am trying to scrape the UESP with the following command:

`mwoffliner --mwUrl=https://en.uesp.net/ --adminEmail=foo@bar.net --verbose --mwWikiPath="/wiki/Main_Page" --mwRestApiPath="/wiki/"`

However it will always failed when it try to download this article: https://en.uesp.net/wiki/Online%3ATreasures%2FHew's_Bane

I could add:  `--articleListToIgnore="https://en.uesp.net/wiki/Online%3ATreasures%2FHew's_Bane"`, however that would mean having to spend about 30 minutes to get to the point of failure each time. And additionally I can only seem to add one article in the list to ignore as if I do the following I get an error:
`--articleListToIgnore="https://en.uesp.net/wiki/Online%3ATreasures%2FHew's_Bane","https://en.uesp.net/wiki/Morrowind:Morrowind"`

**Comment by kelson42** [comment]
404 http responses from article API are already ignored.

**Comment by mai-gh** [comment]
I was able to complete this using the following:
```
docker run -v /my-path:/output:rw ghcr.io/openzim/mwoffliner mwoffliner --adminEmail="foo@bar.net" --customZimDescription="UESPWiki - The Unofficial Elder Scrolls Pages" --customZimFavicon="https://images.uesp.net/favicon-256.png" --customZimTitle="UESPWiki" --filenamePrefix="UESP_en_all" --format="novid:maxi" --mwRestApiPath="/wiki/" --mwUrl="https://en.uesp.net/" --mwWikiPath="/wiki/Main_Page" --osTmpDir="/tmp" --outputDirectory="/output" --publisher="foo" --speed="4" --webp --verbose --resume
```
I mostly based this off of the recipe from https://farm.openzim.org/recipes/elder-scrolls-wiki/config

I have run into another issue about relative links where any link inside is broken:

example:
`http://127.0.0.1:8000/wiki/General:The_Elder_Scrolls` is broken and should be
`http://127.0.0.1:8000/content/uesp_en_all_maxi_2024-01/A/General:The_Elder_Scrolls`

I cheaply got around this by using the [redirector](https://github.com/einaregilsson/Redirector) addon for firefox with a rule:
```
localhost uesp kiwix
Redirect: http://127.0.0.1:8000/wiki/*
to: http://127.0.0.1:8000/uesp_en_all_maxi_2024-01/A/$1
Example: http://127.0.0.1:8000/wiki/Daggerfall:Daedric → http://127.0.0.1:8000/uesp_en_all_maxi_2024-01/A/Daggerfall:Daedric
Applies to: Main window (address bar)
```

this isn't exactly the point of this issue... but is there an easy way to correct this either in the zim file or with kiwix-serve?

also @DataAddict42 if you want I can upload my zim for UESP to archive.org or something

**Comment by DataAddict42** [comment]
> I was able to complete this using the following:
> 
> ```
> docker run -v /my-path:/output:rw ghcr.io/openzim/mwoffliner mwoffliner --adminEmail="foo@bar.net" --customZimDescription="UESPWiki - The Unofficial Elder Scrolls Pages" --customZimFavicon="https://images.uesp.net/favicon-256.png" --customZimTitle="UESPWiki" --filenamePrefix="UESP_en_all" --format="novid:maxi" --mwRestApiPath="/wiki/" --mwUrl="https://en.uesp.net/" --mwWikiPath="/wiki/Main_Page" --osTmpDir="/tmp" --outputDirectory="/output" --publisher="foo" --speed="4" --webp --verbose --resume
> ```
> 
> I mostly based this off of the recipe from https://farm.openzim.org/recipes/elder-scrolls-wiki/config
> 
> I have run into another issue about relative links where any link inside is broken:
> 
> example: `http://127.0.0.1:8000/wiki/General:The_Elder_Scrolls` is broken and should be `http://127.0.0.1:8000/content/uesp_en_all_maxi_2024-01/A/General:The_Elder_Scrolls`
> 
> I cheaply got around this by using the [redirector](https://github.com/einaregilsson/Redirector) addon for firefox with a rule:
> 
> ```
> localhost uesp kiwix
> Redirect: http://127.0.0.1:8000/wiki/*
> to: http://127.0.0.1:8000/uesp_en_all_maxi_2024-01/A/$1
> Example: http://127.0.0.1:8000/wiki/Daggerfall:Daedric → http://127.0.0.1:8000/uesp_en_all_maxi_2024-01/A/Daggerfall:Daedric
> Applies to: Main window (address bar)
> ```
> 
> this isn't exactly the point of this issue... but is there an easy way to correct this either in the zim file or with kiwix-serve?
> 
> also @DataAddict42 if you want I can upload my zim for UESP to archive.org or something

Thanks very much, I will try to get it working myself first as this is a tool that I feel will be really useful to me so it will be beneficial to go through everything myself first. But thanks for the offer

**Comment by DataAddict42** [comment]
@mai-gh I managed to get it downloaded using your command, thanks a lot for your help. I also had the problem of having to use the redirector, hopefully I will be able to find a way to fix the issue at the file level.

**Comment by mai-gh** [comment]
I found a way to edit the `/wiki/` relative urls:
```
zimdump dump --dir=./ud /my-path/UESP_en_all_maxi_2024-01.zim
find ./ud -type f -exec sed -i 's|/wiki/|/content/uesp_en_all_maxi_2024-01/A/|g' {} +
rm -rf ./ud/_exceptions

zimwriterfs --welcome="A/Main_Page" --illustration="M/Illustration_48x48@1" --language="eng" --title="UESPWiki" --name="UESPWiki" --description="USEP Wiki" --longDescription="USEP Wiki" --creator=foo --publisher=foo --threads 4 --verbose ./ud ./uesp_en_all_maxi_2024-01.zim

kiwix-serve --port 8000 --verbose ./uesp_en_all_maxi_2024-01.zim
```

if you run into an error like "too many open files" (I'm on arch linux) I fixed this by following steps from: https://www.howtogeek.com/805629/too-many-open-files-linux/#how-to-permanently-change-the-file-limit

(where they used `4096` i used `8192`)

without extensive testing, this fixed the relative links with the only exception being the homepage doesn't work anymore...
(zimwriterfs wouldn't let me do `--welcome="content/uesp_en_all_maxi_2024-01/A/Main_Page"`

might just be that there is an easy (command line argument) fix with one of the zim / kiwix tools used in this whole process to properly handle the redirects / relative links, but I'm not an expert nor do I have any more time to review the source code etc at the moment. 

I think this is some progress though since a lot of mwoffliner tasks on the farm are failing. I also got the stardew valley wiki to download following these patterns (although with `--mwRestApiPath="/"` instead (which will required a different regex if you go the sed route)).

anyway, cheers @DataAddict42! You just happened to create your first issue when i started down this path for my own reasons..
